### PR TITLE
[data] Ray Data jobs detail table

### PR DESCRIPTION
**Pull request by Zandew**
## Why are these changes needed?

Creates a table under the jobs page to display dataset-level metrics.

The `_StatsActor` now stores dataset metadata like `state`, `progress` and `start/end_time` for each executed dataset, that is directly queried by the new `data_head` dashboard api. This api also makes requests to the prometheus server to get other metrics that are displayed.

Screenshot 2023-10-31 at 4 39 10 PM

cluster env: `jobs-data-overview` to test on workspaces.

## Checks

- [ ] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: dashboard/modules/data/data_head.py [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+MAX_TIME_WINDOW = "1h"
+SAMPLE_RATE = "1s"
+
+
+class PrometheusQuery(Enum):
+    VALUE = ("value", "sum({}) by (dataset)")
+    MAX = (
+        "max",
+        "max_over_time(sum({}) by (dataset)[" + f"{MAX_TIME_WINDOW}:{SAMPLE_RATE}])",
+    )
```
Comment (Zandew) [inline dashboard/modules/data/data_head.py:27]: Is there a simpler way to get the max of a metric?

File: dashboard/modules/data/data_head.py [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+                for metric, queries in DATASET_METRICS.items():
+                    datasets[dataset][metric] = {query.value[0]: 0 for query in queries}
+            # Query dataset metric values from prometheus
+            for metric, queries in DATASET_METRICS.items():
+                for query in queries:
+                    result = await self._query_prometheus(query.value[1].format(metric))
+                    for res in result["data"]["result"]:
+                        dataset, value = res["metric"]["dataset"], res["value"][1]
+                        if dataset in datasets:
+                            datasets[dataset][metric][query.value[0]] = value
```
Comment (Zandew) [inline dashboard/modules/data/data_head.py:62]: We could store these values back in the `_StatsActor` for completed datasets so we don't have to query prometheus for datasets that are done executing?

**Comment by scottsun94** [comment]
Can we detect if people use Ray data in a ray job? If so, I'd like to move this table to the top above the "Task/actor overview" and hide this table when people don't use Ray Data

**Comment by Zandew** [comment]
> Can we detect if people use Ray data in a ray job? If so, I'd like to move this table to the top above the "Task/actor overview" and hide this table when people don't use Ray Data

Yeah we can just hide the table if there aren't any datasets.

**Comment by scottsun94** [comment]
> > Can we detect if people use Ray data in a ray job? If so, I'd like to move this table to the top above the "Task/actor overview" and hide this table when people don't use Ray Data
> 
> Yeah we can just hide the table if there aren't any datasets.

Great! Can you update the screenshot?

File: dashboard/client/src/pages/job/JobDetail.tsx [hunk 3]
```diff
@@ -162,6 +177,17 @@ export const JobDetailChartsPage = () => {
         </Box>
       </CollapsibleSection>
 
+      {data?.datasets && data.datasets.length > 0 && (
```
Comment (scottsun94) [inline dashboard/client/src/pages/job/JobDetail.tsx:180]: We should move it even above the "Tasks/actor overview (beta)". LN 122

File: dashboard/client/src/pages/job/JobDetail.tsx [hunk 4]
```diff
@@ -104,6 +119,17 @@ export const JobDetailChartsPage = () => {
...
+          className={classes.section}
+        >
+          <Section>
+            <DataOverview datasets={data.datasets} />
+          </Section>
+        </CollapsibleSection>
+      )}
+
       <CollapsibleSection
         title="Tasks/actor overview (beta)"
```
Comment (scottsun94) [inline dashboard/client/src/pages/job/JobDetail.tsx:134]: Can we just rename this to "Ray Core Overview"?
cc: @alanwguo

**Review by raulchen — APPROVED** [review]
data changes look good to me

File: dashboard/modules/data/data_head.py (hunk 1 above)
Comment (Zandew) [reply dashboard/modules/data/data_head.py:27]: I think something like `[1h:1s]` in this case means that we're querying the last 1 hour at 1 second intervals.

**Comment by scottsun94** [comment]
LGTM. Thanks!

File: dashboard/BUILD [hunk 5]
```diff
@@ -95,3 +96,10 @@ py_test(
     srcs = ["modules/serve/tests/test_serve_dashboard.py"],
     tags = ["team:serve"],
 )
+
+py_test(
+    name = "test_data_head",
+    size = "small",
+    srcs = ["modules/data/tests/test_data_head.py"],
+    tags = ["team:data"],
+)
```
Comment (can-anyscale) [inline dashboard/BUILD:105]: new line

File: .buildkite/data.rayci.yml [hunk 6]
```diff
@@ -93,6 +93,18 @@ steps:
...
     job_env: forge
 
+  - label: ":database: data: dashboard tests"
+    tags:
+      - python
+      - data
+      - dashboard
+    instance_type: medium
+    commands:
+      - bazel run //ci/ray_ci:test_in_docker -- python/ray/dashboard/... data
```
Comment (can-anyscale) [inline .buildkite/data.rayci.yml:103]: missing `--build-name datanbuild`

also datanbuild is a build with nightly pyarrow, there are also data6build and data12build which pins pyarrow to 6.* and 12.* version; not sure which one you really want just fyi

File: .buildkite/data.rayci.yml [hunk 7]
```diff
@@ -93,6 +93,18 @@ steps:
...
+        --parallelism-per-worker 3
+    depends_on: datanbuild
```
Comment (can-anyscale) [inline .buildkite/data.rayci.yml:105]: whatever build you choose this needs to upgrade to that build name

File: dashboard/client/src/pages/job/JobDetail.tsx [hunk 8]
```diff
@@ -104,8 +119,19 @@ export const JobDetailChartsPage = () => {
...
+        >
+          <Section>
+            <DataOverview datasets={data.datasets} />
+          </Section>
+        </CollapsibleSection>
+      )}
+
       <CollapsibleSection
-        title="Tasks/actor overview (beta)"
+        title="Ray Core Overview"
```
Comment (scottjlee) [inline dashboard/client/src/pages/job/JobDetail.tsx:134]: is this related to the other code changes / intended?

File: dashboard/modules/data/data_head.py (hunk 2 above)
Comment (scottjlee) [reply dashboard/modules/data/data_head.py:62]: yeah, let's do this in a future PR (feel free to leave a `TODO` comment or open an issue)

File: dashboard/modules/data/data_head.py [hunk 9]
```diff
@@ -0,0 +1,95 @@
...
+            for dataset in datasets:
+                for metric, queries in DATASET_METRICS.items():
+                    datasets[dataset][metric] = {query.value[0]: 0 for query in queries}
+            # Query dataset metric values from prometheus
+            for metric, queries in DATASET_METRICS.items():
+                for query in queries:
+                    result = await self._query_prometheus(query.value[1].format(metric))
+                    for res in result["data"]["result"]:
+                        dataset, value = res["metric"]["dataset"], res["value"][1]
+                        if dataset in datasets:
```
Comment (scottjlee) [inline dashboard/modules/data/data_head.py:61]: is there any case where this would not be true? if so, would we want to also include the dataset in the results?

**Review by scottjlee — APPROVED** [review]
nice work!

File: dashboard/client/src/pages/job/JobDetail.tsx (hunk 8 above)
Comment (Zandew) [reply dashboard/client/src/pages/job/JobDetail.tsx:134]: Just a change @scottsun94 wanted, to make a clear distinction between what table ray data/core users should be looking at

File: dashboard/modules/data/data_head.py (hunk 9 above)
Comment (Zandew) [reply dashboard/modules/data/data_head.py:61]: If you don't restart the prometheus server, this can have extra datasets across different jobs

**Review by c21 — APPROVED** [review]
LGTM for data side.

**Comment by c21** [comment]
Hi @alanwguo - do you wanna do a final review on dashboard side? Thanks.

**Review by alanwguo — APPROVED** [review]

**Review by can-anyscale — APPROVED** [review]
